src/main/java/com/thirdcc/webapp/service/impl/UserCCInfoServiceImpl.java

```diff
@@ -91,6 +99,83 @@ public UserCCInfoDTO save(UserCCInfoDTO userCCInfoDTO) {
+        List<String> yearSessionWithCCFamilyRole = userCCInfoDTOList.stream()
+            .map(UserCCInfoDTO::getYearSession)
+            .collect(Collectors.toList());
+        List<UserCCInfoDTO> fullUserCCInfoDTOList = new ArrayList<>();
+        userUniInfoService.getUserUniInfoByUserId(userId)
+            .ifPresent(userUniInfoDTO -> {
+                String currentYearSession = YearSessionUtils.getCurrentYearSession();
+                String intakeYearSession = userUniInfoDTO.getYearSession();
+                String secondYearSession = YearSessionUtils.addYearSessionWithSemester(intakeYearSession, 2);
+                if (currentYearSession.compareTo(intakeYearSession) > -1) {
+                    if (!yearSessionWithCCFamilyRole.contains(intakeYearSession)) {
+                        fullUserCCInfoDTOList.add(constructCustomUserCCInfo(intakeYearSession, FishLevel.JUNIOR_FISH, userId));
+                    }
+                    else {
+                        UserCCInfoDTO userCCInfoDTO = userCCInfoDTOList.stream()
+                            .filter(userCCInfoDTO1 -> userCCInfoDTO1.getYearSession().equals(intakeYearSession))
+                            .findFirst()
+                            .get();
+                        userCCInfoDTO.setFishLevel(FishLevel.JUNIOR_FISH);
+                        fullUserCCInfoDTOList.add(userCCInfoDTO);
+                    }
+                }
+                if (currentYearSession.compareTo(secondYearSession) > -1) {
+                    if (!yearSessionWithCCFamilyRole.contains(secondYearSession)) {
+                        fullUserCCInfoDTOList.add(constructCustomUserCCInfo(secondYearSession, FishLevel.SENIOR_FISH, userId));
+                    }
+                    else {
+                        UserCCInfoDTO userCCInfoDTO = userCCInfoDTOList.stream()
+                            .filter(userCCInfoDTO1 -> userCCInfoDTO1.getYearSession().equals(secondYearSession))
+                            .findFirst()
+                            .get();
+                        userCCInfoDTO.setFishLevel(FishLevel.SENIOR_FISH);
+                        fullUserCCInfoDTOList.add(userCCInfoDTO);
+                    }
+                }
+                if (currentYearSession.compareTo(secondYearSession) > 0) {
+                    String thirdYearSession = YearSessionUtils.addYearSessionWithSemester(secondYearSession, 2) + " - Present";
+                    fullUserCCInfoDTOList.add(constructCustomUserCCInfo(thirdYearSession, FishLevel.ELDER_FISH, userId));
+                }
+                userCCInfoDTOList.stream()
+                    .filter(userCCInfoDTO1 -> userCCInfoDTO1.getYearSession().compareTo(secondYearSession) > 0)
+                    .forEach(fullUserCCInfoDTOList::add);
+            });
+        fullUserCCInfoDTOList.sort(Comparator.comparing(UserCCInfoDTO::getYearSession));
```

Comment:
This code is too hard to read and maintain.
Suggest you use a loop.
- increment yearSession by 1 every time
- determine fish level based on loop index `i`
- then no need to sort using sort function also, since it will guarantee to be in ascending Year Session

Reply:
@kachun333 
Can I separate the if into 3 diff methods instead of using loop?

What I think is Loop performance is slower

Reply:
resolved in d5253a35de87edf18588bc0c2bae3356b925f969